--- linux/drivers/char/rtc.c.orig	Mon Mar 20 12:13:57 2000
+++ linux/drivers/char/rtc.c	Mon Mar 20 12:23:10 2000
@@ -575,13 +575,14 @@
 	if (!(ctrl & RTC_DM_BINARY) || RTC_ALWAYS_BCD)
 		BCD_TO_BIN(year);       /* This should never happen... */
 	
-	if (year > 10 && year < 44) {
-		epoch = 1980;
-		guess = "ARC console";
-	} else if (year < 96) {
-		epoch = 1952;
-		guess = "Digital UNIX";
-	}
+	if (year >= 19)
+		if (year < 44) {
+			epoch = 1980;
+			guess = "ARC console";
+		} else if (year < 96) {
+			epoch = 1952;
+			guess = "Tru64 UNIX";
+		}
 	if (guess)
 		printk("rtc: %s epoch (%lu) detected\n", guess, epoch);
 #endif

