*** axpPci.c    2000/04/10 13:13:36     1.1.1.1
--- axpPci.c    2000/07/06 08:48:56
***************
*** 101,108 ****
        CARD32 val = 0xffffffff;
 
        bus = PCI_BUS_FROM_TAG(tag);
        dfn = PCI_DFN_FROM_TAG(tag);
!
        syscall(__NR_pciconfig_read, bus, dfn, off, 4, &val);
        return(val);
  }
--- 101,114 ----
        CARD32 val = 0xffffffff;
 
        bus = PCI_BUS_FROM_TAG(tag);
+       /*
+        * Workaround for kernel bug
+        * triggered when probing for PCI devices
+        */
+       if (bus >= pciNumBuses)
+           return (val);
        dfn = PCI_DFN_FROM_TAG(tag);
!
        syscall(__NR_pciconfig_read, bus, dfn, off, 4, &val);
        return(val);
  }                                                                              


