--- hatman-0.5.0.1/util/error.cc	Sun Feb 28 23:20:52 1999
+++ hatman-0.5.0.1-alpha/util/error.cc	Mon Jun 21 19:54:17 1999
@@ -31,9 +31,8 @@
 
 char* errStr = "UNDEFINED ERROR";
 
-static int printError(FILE* stream, const struct printf_info* info, va_list* app)
+static int printError(FILE* stream, const struct printf_info* info, const void *const *args)
 {
- (void)info; (void)app;
  if(!errStr)
   {
    fprintf(stream, "(no error)");
@@ -49,9 +48,9 @@
  return 0;
 }
 
-static int printErrnoString(FILE* stream, const struct printf_info* info, va_list* app)
+static int printErrnoString(FILE* stream, const struct printf_info* info, const void *const *args)
 {
- (void)info; (void)app;
+ (void)info;
  int i = errno;
  fprintf(stream, "%s", strerror(i));
  return(strlen(strerror(i)));

--wRRV7LY7NUeQGEoC--


--  
To UNSUBSCRIBE, email to debian-alpha-request@lists.debian.org
with a subject of "unsubscribe". Trouble? Contact listmaster@lists.debian.org


